### ISSUE TITLE
[RFC]: Application-Directed Prefix Checkpoints for Mamba / Hybrid Prefix Caching

### ISSUE THREAD

**Issue by nicholaskh-ai** [vllm-project/vllm#55697]
## Motivation.

> **TL;DR**: We propose an application-directed prefix checkpointing mechanism for hybrid SSM/Attention models (Qwen3.5, Nemotron-4, Jamba) in vLLM. By allowing applications to declare semantic prefix boundaries via an input-side control marker (`<|mamba_checkpoint|>`), vLLM coordinates Producer-Consumer prefill scheduling to eliminate redundant prefix computations across requests.
>
> **Key Production Results (Qwen3.5 35B on NVIDIA L40S)**:
> - 🚀 **+110.2% Throughput (2.1x Speedup)**: End-to-end serving QPS increased from **5.9 QPS to 12.4 QPS**.
> - 🎯 **Zero Accuracy Loss (100% Numerical Parity)**: Exact logit matching with eager baseline ($< 1.2 \times 10^{-5}$ max diff in bfloat16).
> - ⚡ **48.6% TTFT Reduction**: Prefill latency dropped from 482 ms to 248 ms.
> - 📉 **53.4% GPU FLOPs Reduction**: Prefill computation cut from ~11,500 tokens/batch to ~5,358 tokens/batch.
> - 🏎️ **Up to 7.03x Kernel Speedup**: Batched phased GDN prefill kernel vs. legacy loop on L40S.
>
> **Evaluation Dataset & Workload Profile**:
> - **Structure**: 1-to-9 Multimodal Pairwise Matching benchmark (1 shared anchor item vs. 9 distinct candidate items, evaluated in 10-request groups).
> - **Prefix Anatomy**: Strictly identical shared prefix of $p \approx 624$ tokens post-expansion (structured text attributes + 1 high-resolution main image).
> - **Suffix Anatomy**: Diverging, variable-length suffixes ranging from 350 to 650 tokens (candidate text specifications + 1 candidate image). Total prompt lengths vary between 980 ~ 1,300 tokens.

#### Problem: The Memory & Boundary Dilemma in Recurrent Prefix Caching
Existing prefix caching in vLLM operates seamlessly for Transformer/Full-Attention layers where each token position independently retains a fine-grained $K/V$ vector. In recurrent models like Mamba/GDN, however, hidden states are sequential, stateful, and **substantially larger per checkpoint** (e.g., full $[H, D_k, D_v]$ state matrices per layer, ~512 KB per layer). Due to this memory footprint, recurrent architectures cannot afford dense token-wise caching and can realistically retain only **1 or 2 discrete state snapshots (checkpoints)** per sequence.

The core dilemma is that **from an engine-internal perspective, it is virtually impossible to heuristically predict where that single, high-value checkpoint boundary should be placed**. Geometry-based interval chopping frequently misses the actual shared premise or prematurely cuts across context. Conversely, **from the business application's perspective, this boundary is immediately obvious and trivial to locate** (e.g., precisely at the junction where a shared source item, document context, or system premise ends).

#### Real-World Workload: 1-to-$N$ Multimodal Item Matching
In high-throughput item matching pipelines (such as product deduplication and candidate item pair matching), a critical production task is **1-to-$N$ Candidate Item Matching**:
* **Input**: Given 1 anchor source item (`Item 1`), the system compares it against 9 candidate items (`Candidate 0` to `Candidate 8`), forming **10 item-pair inference requests** (1 Producer + 9 Consumers).
* **Multimodal Data**:
  - Each item contains rich textual metadata: item title, brand, manufacturer, and structured specifications.
  - Each item contains **1 high-resolution main product image** (`main image`).
  - Thus, every single request contains 2 images: `[Item 1 image, Candidate i image]`.
* **Output**: A sequence classification or scoring probability (logits) predicting whether `Item 1` and `Candidate i` represent the exact same physical product item.

#### The Variable-Suffix Geometry Problem
Each candidate item has distinct titles, varying attribute lengths, and different image aspect ratios/resolutions. Consequently, **the total prompt lengths across the 10 requests are inherently variable**:
```text
Request 0 (Producer): [Item 1 Text + Image 1] <marker> [Candidate 0 Text + Image 2_0] -> Total:  982 tokens (Suffix: 358 toks)
Request 1 (Consumer): [Item 1 Text + Image 1] <marker> [Candidate 1 Text + Image 2_1] -> Total: 1248 tokens (Suffix: 624 toks)
Request 2 (Consumer): [Item 1 Text + Image 1] <marker> [Candidate 2 Text + Image 2_2] -> Total: 1056 tokens (Suffix: 432 toks)
...
Request 9 (Consumer): [Item 1 Text + Image 1] <marker> [Candidate 9 Text + Image 2_9] -> Total: 1169 tokens (Suffix: 545 toks)
```

* **The Shared Invariant**: The prefix `[Item 1 Text + Image 1]` is **strictly identical** across all 10 requests (e.g., exactly $p = 624$ tokens post-multimodal expansion).
* **Full Attention Behavior**: vLLM's `BlockPool` easily aliases physical KV blocks for `Item 1` because token-wise attention vectors are independent of suffix length.
* **Mamba / GDN Dilemma**: Because Mamba recursively mutates its recurrent state in place, if all 10 requests are scheduled without explicit coordination:
  1. The engine cannot deduce where the common premise ends.
  2. Each request runs `Item 1` from scratch, wasting $9 \times 624 = 5,616$ redundant tokens per 10 requests.
  3. If requests attempt uncoordinated same-step execution, they either overwrite the shared live recurrent state (corrupting classification scores) or miss the cache entirely (`local_hit = 0`).

#### Why the Boundary MUST Be Declared by the Application Layer
1. **Multimodal Dynamic Token Expansion**: In models like Qwen2.5-VL and Qwen3-VL, image tokens expand dynamically based on aspect ratio and patch merging. The engine cannot know through geometry alone whether an image belongs to `Item 1` or `Candidate 2`.
2. **Semantic Awareness vs. Black-Box Tokens**: The business application knows with 100% certainty where `Item 1` ends and `Candidate` begins (right after `Item 1`'s image placeholder). The inference engine, seeing only a flat array of token IDs, has no semantic context.
3. **Severe State Memory Budget (1-2 Snapshots Max)**: Because a sequence can realistically afford only 1 or 2 state checkpoints, squandering this budget on an arbitrary heuristic interval wastes memory. Guiding the checkpoint from the business layer ensures that the single allocated state snapshot aligns 100% with real traffic reuse patterns.

---

#### 1. Input-Side Control Token
* **Master Switch**: `--enable-mamba-checkpoint` (`enable_mamba_checkpoint: bool = False`, disabled by default for zero regression risk).
* **Default Token**: `--mamba-checkpoint-token` defaults to `<|mamba_checkpoint|>`.
* **Zero Output Side-Effects**: The marker is registered as an additional special token in both framework and model tokenizers. After position extraction, the token is stripped from `prompt_token_ids` before request hashing, validation, or execution. The model never generates the token and downstream output logits are completely unaffected.
* **Multimodal-Aware**: For vision-language models, the checkpoint position $p$ is calculated **after** multimodal placeholder expansion, ensuring $p$ represents the exact sequence position in the LLM backbone.

#### 2. Chained Block Hashing
The checkpoint identifier is embedded within vLLM's existing chained block hasher:
```text
Hash_k = SHA256(Hash_{k-1} + Tokens_k + ExtraKeys_k)
```
Where `ExtraKeys` includes the image SHA256, placeholder offsets, and LoRA identity.
Because $p$ is aligned to `prefix_match_unit` (e.g., 16), the checkpoint falls exactly on block index $k = p / 16 - 1$.
The Mamba state is registered under:
```python
Key = make_block_hash_with_group_id(Hash_k, group_id=1)
```

#### 3. Scheduler State Machine (Producer-Consumer Coordination)
```text
User Prompt (Item Matching Pair):
"Item 1: title, brand, attrs... <image> <|mamba_checkpoint|> Candidate 2: title... <image>"
       │
       ▼
┌────────────────────────────────────────────────────────┐
│ 1. InputProcessor (Frontend / Renderer)                │
│    - Detect <|mamba_checkpoint|> token ID              │
│    - Compute post-multimodal expansion position p      │
│    - Align p to prefix_match_unit (e.g., 16 / 32)      │
│    - Strip marker from input_ids (0 impact on output)  │
│    - Store request.mamba_checkpoint_position = p       │
└────────────────────────────────┬───────────────────────┘
                                 │
                                 ▼
┌────────────────────────────────────────────────────────┐
│ 2. Scheduler & KV Cache Coordinator                    │
│    - Request 0 (Producer):                             │
│        - Clips prefill chunk at p = 624                │
│        - Caches Mamba state block at index 38          │
│        - Marks checkpoint hash as UNREADY              │
│    - Request 1..9 (Consumers with variable suffixes):  │
│        - Detect checkpoint hash is UNREADY             │
│        - Automatically deferred to step_skipped_waiting│
└────────────────────────────────┬───────────────────────┘
                                 │ Forward Step 1
                                 ▼
┌────────────────────────────────────────────────────────┐
│ 3. Forward Step 1 (Producer Prefix Only)               │
│    - Producer computes tokens 0..624                   │
│    - Mamba state at 624 committed to CacheBlock        │
│    - mark_checkpoint_ready(hash_38) called             │
└────────────────────────────────┬───────────────────────┘
                                 │ Forward Step 2
                                 ▼
┌────────────────────────────────────────────────────────┐
│ 4. Forward Step 2 (Batched Consumers)                  │
│    - Consumers 1..9 woken up                           │
│    - find_longest_cache_hit -> local_hit = 624         │
│    - Skip tokens 0..624, read shared checkpoint state  │
│    - Batch compute all 10 variable suffixes in parallel│
└────────────────────────────────────────────────────────┘
```
* **Producer Truncation**: When `cache_config.mamba_cache_mode == "align"`, `_mamba_block_aligned_split` clamps the Producer's scheduled tokens to $p$.
* **Unready Lock**: The Producer marks `block_pool.mark_block_hash_unready(checkpoint_hash)`.
* **Automatic Consumer Deferral**: Subsequent requests sharing the same prefix detect `has_unready_checkpoint(req) == True` and are moved to `step_skipped_waiting`.
* **Atomic Wakeup**: Once the Producer's step completes, `mark_checkpoint_ready` releases the lock. On the next step, all Consumers resume, achieve `local_hit = p`, and execute their suffixes in parallel.

#### 4. Memory Efficiency & State Isolation
* **Zero Extra Physical Blocks**: For tokens $0 \dots p-1$, Mamba assigns `NullBlock(block_id=0)` pointers in software. Only 1 physical block (e.g. 512 KB per layer) is allocated for the checkpoint state at position $p$.
* **Read-Only Sharing with CoW**: Consumers read the Producer's checkpoint state into their private hidden state registers. Write destinations are strictly private per request, ensuring 100% numerical isolation.

---

### Feedback Period.

Two weeks after posting. The main review questions are:
1. Is `--enable-mamba-checkpoint` and `--mamba-checkpoint-token` the preferred CLI / engine interface for application-directed prefix boundaries?
2. Does the Producer-Consumer pending/wakeup state machine in `scheduler.py` cleanly integrate with upcoming disaggregated prefill/decode plans?
3. What is the community consensus on fusing Producer and Consumer phases into a single persistent kernel to eliminate HBM traffic?

---

### CC List.

@ZJY0516 @roikoren755 @yewentao256 @youkaichao @WoosukKwon @benchislett

---

#### Relationship to Existing Work
- **Strictly Complementary to #52959** (*Internal State Checkpoints for Mamba Align Mode* by @ZJY0516):
  - **#52959** solves **intra-request, single-forward kernel chunking** (deriving intermediate state snapshots at fixed mathematical intervals $B$ inside one forward pass to avoid splitting long prefills).
  - **This RFC** solves **inter-request, semantic-level 1-to-$N$ prefix caching and scheduling** (allowing business applications to declare domain-level shared premises across requests with variable suffixes).
  - When RFC #52959's internal checkpoint kernel is available, a Producer does not even need a separate forward pass to commit its checkpoint; it can compute the prefix and its own suffix in a single forward pass, while emitting the state checkpoint for waiting Consumers.
- **Directly addresses the open question in #45702** (*Partial Cache Hits for Hybrid Models* by @ZJY0516):
  - Section *"Open Design Question: Same-Step Reuse"* specifically noted: *"Whether to implement same-step reuse or same-step recomputation is an open question."*
  - This RFC provides the working, verified answer to that question through scheduler Producer-Consumer coordination.
- **Orthogonal to #52817** (*Hybrid SSM models & last-block-replay with SpecDec/APC* by @roikoren755):
  - Focuses on drafter intermediate state replay for EAGLE/MTP.

#### Future Kernel Optimization: Fusing Producer and Consumer Operators
Currently, the batched grouped GDN prefill executes as two sequential kernel phases (`run_phase` for Producer followed by `run_phase` for Consumers) within the same CUDA stream. While this completely eliminates redundant token FLOPs, the intermediate Mamba recurrent state at position $p$ (~512 KB per layer) is written to GPU HBM in Phase 1 and re-read by all $N$ Consumers from HBM in Phase 2.
As a planned follow-up, we are investigating a **fused single-operator kernel** (persistent kernel or tree-scan GDN) that keeps the Producer's boundary hidden state resident directly in on-chip SRAM / L2 Cache:
- Eliminates $N \times 512\text{ KB}$ of global HBM read bandwidth per layer per batch.
- Halves kernel launch overhead (e.g. from 96 launches to 48 launches for a 24-layer model), further reducing CPU dispatch overhead on lightweight models like Qwen3.5-0.8B.

#### Upstream Contribution Checklist
- [x] Unit tests covering scheduler and metadata builder (`tests/v1/core/test_mamba_checkpoint_scheduler.py`, `tests/v1/attention/test_gdn_metadata_builder.py`)
- [x] Kernel mock correctness tests and L40S microbenchmarks (`tests/kernels/mamba/`, `benchmarks/kernels/benchmark_grouped_gdn_prefill.py`)
- [ ] Upstream community review and feedback
- [x] Open implementation PRs linked to this RFC (#55873, #55875, #55876)

---

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by siva23367** [comment]
I want to work on this issue , could you please assign this issue to me @nicholaskh-ai

**Comment by frozenbugs** [comment]
We have similar issues and this feature can benefit us a lot.

**Comment by nicholaskh-ai** [comment]
Update: The implementation of this RFC has been split into 3 progressive, layered PRs and submitted for upstream review:

1. **Part 1 (Frontend & Input Processing)**: https://github.com/vllm-project/vllm/pull/55873
   - `[Feature][Config] Support application-directed Mamba checkpoint token and input processing`
   - Covers CLI/Engine options (`--enable-mamba-checkpoint`, `--mamba-checkpoint-token`), `<|mamba_checkpoint|>` token dynamic registration in HF renderer, InputProcessor position extraction, and multimodal placeholder offset calibration.

2. **Part 2 (Scheduler Coordination & State Machine)**: https://github.com/vllm-project/vllm/pull/55875
   - `[Feature][Core] Prefix checkpoint coordination and same-step pairing in V1 scheduler`
   - Covers Producer checkpoint boundary chunk truncation, BlockPool unready hash lock, Consumer defer/wakeup state machine, and same-step Producer/Consumer prefix block inheritance.

3. **Part 3 (Worker & Kernel Grouped Prefill Execution)**: https://github.com/vllm-project/vllm/pull/55876
   - `[Feature][Mamba] Batched two-phase grouped prefill execution for hybrid GDN/Mamba (+7.58x)`
   - Covers two-phase batched kernel execution in GDN layer (`run_phase` for Producer followed by Consumer phase with `index_select` / `index_copy_`), grouped metadata builder, and reproducible benchmark on NVIDIA L40S demonstrating up to **7.58x speedup**.

Feedback and reviews are warmly welcomed!

**Comment by nicholaskh-ai** [comment]
CC @mgoin @jeejeelee @WoosukKwon @yewentao256 @ZJY0516

Hi maintainers, following up on this RFC, we have broken down the implementation into 3 self-contained, reviewable PRs (#55873, #55875, #55876) with comprehensive unit tests and verified NVIDIA L40S microbenchmarks showing up to **7.58x speedup**.

Whenever you have a chance, we would greatly appreciate it if you could take a look and add the `ready` label to kick off CI. Thank you!

**Comment by nicholaskh-ai** [comment]
CC @WoosukKwon @mgoin @jeejeelee @ZJY0516

Hi maintainers,

Following up on RFC #55697: as noted by @yewentao256 in PR #55876, we would like to align with you on the architectural RFC before moving forward with the PR stack.

### Motivation & Workload Characteristics
In hybrid SSM/Attention models (such as Qwen3.5, Nemotron-4), hidden states are dense recurrent matrices (~512 KB per layer). Due to this memory footprint, caching dense per-token states like Transformers is prohibitive; sequences can realistically afford only 1–2 discrete state checkpoints.

In high-throughput 1-to-N structured queries (such as pairwise product deduplication, candidate matching, or document premise queries with variable-length candidate suffixes):
- An identical multimodal premise (e.g., text attributes + high-resolution product image, ~624 tokens) is shared across 10 concurrent requests.
- Suffixes diverge in length (350–650 tokens).
- **The Core Dilemma**: From an engine-internal perspective, it is virtually impossible to heuristically predict where this single high-value boundary should be placed without cutting across atomic multimodal image spans. From the business application's perspective, this boundary is semantic, obvious, and trivial to mark (`<|mamba_checkpoint|>`).

### Verified Production Results (Qwen3.5 35B on NVIDIA L40S)
- **Serving Throughput**: Baseline (eager vLLM) **5.9 QPS -> 13.8 QPS** (+133.9% throughput, 2.34x speedup) on native Model Runner V2 (MRv2).
- **Prefill Latency (TTFT)**: Cut by **48.6%** (482 ms -> 248 ms).
- **GPU FLOPs**: Prefill compute cut from ~11,500 tokens/batch to ~5,358 tokens/batch.
- **Accuracy Parity**:
  - `lm_eval` GSM8K (5-shot): exact match parity (**0.8522 vs 0.8522**, 0.00% accuracy loss).
  - 3,428 production request pairs: **0.9991 Pearson correlation**, **99.94% prediction consistency**, **0.00% significant diff (`>= 0.1`)**.

### System Complexity & Isolation
To address maintainability concerns:
1. **Zero Regression Risk**: All components are feature-gated via `--enable-mamba-checkpoint` (`False` by default). Standard Transformer and un-gated Mamba serving paths are completely untouched.
2. **Native MRv2 Adaptation**: The kernel execution and state inheritance (`mamba_hybrid.py`, `gdn_attn.py`) run natively on Model Runner V2 with zero legacy MRv1 code.
3. **Clean 3-Layer Modularity**:
   - Layer 1 (PR #55873): Input-side marker detection & stripping.
   - Layer 2 (PR #55875): Core Scheduler boundary chunk splitting & BlockPool unready lock.
   - Layer 3 (PR #55876): MRv2 model state inheritance & batched GDN execution.

We would greatly appreciate your guidance on whether this direction fits vLLM's long-term hybrid SSM roadmap, and how you would prefer we shape this feature to minimize maintenance overhead for the community. Thank you!
